DELL Platform 2.0 API Infra and Reboot Reason support in Z9100 & S6100

sridhar-ravindran opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Added platform 2.0 API support infra for Dell platform s6100 & z9100
**- How I did it**
https://github.com/Azure/SONiC/blob/540d890d3acbc9e424dac75f0f671cbb352f0a52/doc/platform_api/new_platform_api.md  
Adhered to the latest solution proposed.

This code has dependancy on 
https://github.com/Azure/sonic-buildimage/pull/3059
**- How to verify it**
Added a stub code in show/main.py and verified show reboot-cause command.
UT logs attached with more details.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
DELLEMC : S6100,Z9100 Platform 2.0 API Infra Support

**- A picture of a cute animal (not mandatory but encouraged)**

sridhar-ravindran commented:
[UT_Logs_For_Reboot_Reason.txt](https://github.com/Azure/sonic-buildimage/files/3315741/UT_Logs_For_Reboot_Reason.txt)

jleveque commented on platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/chassis.py:
```diff
@@ -0,0 +1,79 @@
+        else:
+            if (reset_reason in self.reset_reason_dict):
+                return self.reset_reason_dict[reset_reason]
+        return ((),)
```
What is this default return value tuple? This function should only ever return one of the `REBOOT_CAUSE_*` constants. If it is only possible to reach this line if the previous reboot was not caused by hardware, then it should return `REBOOT_CAUSE_NON_HARDWARE`, otherwise there is also `REBOOT_CAUSE_HARDWARE_OTHER` which is to be used as a catch-all for reasons without a related constant defined. This can be used as a temporary stop-gap until we decide whether to add more constants for other hardware-related reboot causes.

sridhar-ravindran replied:
Hi @jleveque 
In S6100 & Z9100, All the reboot reasons are retrieved only from reset_reason and power_on_reason   SMF registers. Added the final return statement to handle the negative scenario in case registers returned incorrect data

jleveque replied:
See my new comment below.

sridhar-ravindran commented:
Hi @jleveque 

Thanks for the review comments. Will address it and upload soon.

Thanks
Sridhar.R

sridhar-ravindran commented on dockers/docker-platform-monitor/start.sh.j2:
```diff
@@ -37,6 +37,12 @@ if [ -e /usr/share/sonic/platform/fancontrol ]; then
 fi
 
+# If this platform has a python api wheel file, install it
+if [ -e /usr/share/sonic/platform/sonic_platform-1.0-py2-none-any.whl ]; then
```
Hi @jleveque 
updated the start.sh script with the logic you provided. Verified the cases as well.

stephenxs replied:
@jleveque @sridhar-ravindran 
On other platforms, the sonic-platform whl may have different names, which fails this piece of code. It doesn't impact the normal function but generates a piece of log like:

    INFO pmon#supervisord: start.sh Error: Unable to locate /usr/share/sonic/platform/sonic_platform-1.0-py2-none-any.whl

To resolve this I have some suggestions:
1. Make this piece code vendor-specific.
2. Make `SONIC_PLATFORM_WHEEL` and `pip show sonic-platform` vendor-specific
3. remove the message "Error: Unable to locate ${SONIC_PLATFORM_WHEEL}"

jleveque replied:
@stephenxs: I think it would be great if we could standardize this across all platforms. I can understand that different vendors may want to increment the version of their package, but I don't see a reason why anything else in the package name must be different. Maybe we could simply find the package by looking for a package matching `/usr/share/sonic/platform/sonic_platform-*-py2-none-any.whl`?

sridhar-ravindran commented on platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/chassis.py:
```diff
@@ -0,0 +1,79 @@
+    MAILBOX_DIR = HWMON_DIR + HWMON_NODE
+
+    reset_reason_dict = {}
+    reset_reason_dict[11] = [(ChassisBase.REBOOT_CAUSE_POWER_LOSS)]
```
Hi @jleveque
I still kept it as ChassisBase. Hope it is fine.

sridhar-ravindran commented:
Hi @jleveque 
**Platform 2.0 Reboot reason has dependancy on :- https://github.com/Azure/sonic-buildimage/pull/3059**

jleveque commented on platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/chassis.py:
```diff
@@ -0,0 +1,82 @@
+            if (reset_reason in self.reset_reason_dict):
+                return self.reset_reason_dict[reset_reason]
+
+        return [(ChassisBase.REBOOT_CAUSE_HARDWARE_OTHER, "Invalid Reason")]
```
I understand you're using a tuple in an attempt to provide a reason with REBOOT_CAUSE_HARDWARE_OTHER. however, I'm not a fan of returning a tuple here in this special case and a string for every other case. I think this function needs to consistently return the same data type.

Would it be better if we added a new constant for `REBOOT_CAUSE_HARDWARE_INVALID_REASON`?

sridhar-ravindran replied:
Hi @jleveque 
Agreed REBOOT_CAUSE_HARDWARE_INVALID_REASON is redundant. 
Removed the special case string. We will fallback to REBOOT_CAUSE_HARDWARE_OTHER in a negative case scenario. Ideally it will never hit that statement,

Thanks
Sridhar.R

jleveque replied:
Can you also please change the return type to just the string, not in a list? The same applies to the values in `reset_reason_dict` and `power_reason_dict`; the function should only ever return one of the predefined strings.

sridhar-ravindran replied:
Hi @jleveque 
Reverted back to returning as tuple, 
In negative case we return hardware_other with invalid reason

Test Example 

root@sonic:~# show reboot-cause
REBOOT REASON : [('hardware_other', 'Invalid Reason')]               <<<<<<<<<<<<<<<<<<<<
User issued 'reboot' command [User: admin, Time: Wed Jun 26 02:47:21 UTC 2019]

sridhar-ravindran commented:
Hi @jleveque ,
I have addressed all the review comments. Could you please check and merge this
along with https://github.com/Azure/sonic-buildimage/pull/3059
Thanks
Sridhar.R

jleveque commented on platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/chassis.py:
```diff
@@ -0,0 +1,82 @@
+            if (reset_reason in self.reset_reason_dict):
+                return self.reset_reason_dict[reset_reason]
+
+        return [(ChassisBase.REBOOT_CAUSE_HARDWARE_OTHER, "Invalid Reason")]
```
I'm not sure if returning the tuple inside a list will work here. What is the reasoning here for using a list?

sridhar-ravindran replied:
Hi @jleveque 
Returned the list for readability only, No specific reason. Removed the list and returning it as tuple.

jleveque commented on platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/chassis.py:
```diff
@@ -0,0 +1,82 @@
+    MAILBOX_DIR = HWMON_DIR + HWMON_NODE
+
+    reset_reason_dict = {}
+    reset_reason_dict[11] = (ChassisBase.REBOOT_CAUSE_POWER_LOSS)
```
Parenthesizing these here is unnecessary and a bit confusing since they are individual values, not tuples. I suggest removing the parentheses from around these values and changing lines 74-79 as follows to make it more clear:

```
        if (reset_reason == 11):
            if (power_reason in self.power_reason_dict):
                return (self.power_reason_dict[power_reason], None)
        else:
            if (reset_reason in self.reset_reason_dict):
                return (self.reset_reason_dict[reset_reason], None)
```

sridhar-ravindran replied:
Hi @jleveque 

Understood. Removed the paranthesis and returned None at relevant places..

thanks
Sridhar.R

jleveque commented on platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/chassis.py:
```diff
@@ -0,0 +1,82 @@
+    power_reason_dict[11] = ChassisBase.REBOOT_CAUSE_POWER_LOSS
+    power_reason_dict[22] = ChassisBase.REBOOT_CAUSE_THERMAL_OVERLOAD_CPU
+    power_reason_dict[33] = ChassisBase.REBOOT_CAUSE_THERMAL_OVERLOAD_ASIC
+    power_reason_dict[44] = ChassisBase.REBOOT_CAUSE_INSUFFICIENT_FAN
```
FYI, yesterday I changed this constant name to `REBOOT_CAUSE_INSUFFICIENT_FAN_SPEED`. Please update accordingly.

sridhar-ravindran replied:
Hi @jleveque 
Updated the constant name.

jleveque commented:
Please resolve conflicts.

sridhar-ravindran commented:
Hi @jleveque 
Resolved the conflicts.

sridhar-ravindran commented:
Hi @jleveque

I have resolved the conflicts. Request you to merge the changes

Thanks
Sridhar.R

jleveque commented:
@sridhar-ravindran: I will merge once all check builds pass.

jleveque commented:
Retest vs please

yxieca commented:
This commit cannot be cherry-picked cleanly into 201811 branch.

sridhar-ravindran commented:
> This commit cannot be cherry-picked cleanly into 201811 branch.

Hi Ying,
Fixed merge issues and raise new pull request in 201811 branch already
https://github.com/Azure/sonic-buildimage/pull/3147
